torch/distributed/_tensor/dispatch.py

```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+        # 3. if the return type is List[Tensor], return a list of empty
+        # tensors with correct dtype.
+
+        def _default_tensor(spec: Optional[DTensorSpec]):
+            assert spec is not None, (
+                f"expect DTensorSpec object but got {spec}"
+            )
+            if spec.tensor_meta is not None:
+                shape = spec.tensor_meta.shape
+                dtype = spec.tensor_meta.dtype
+                if len(shape) == 0:
+                    # scalar tensor
+                    return torch.zeros(
+                        (),
+                        dtype=dtype,
+                    )
+                else:
+                    # non-scalar tensor
+                    return torch.tensor(
+                        [],
+                        dtype=dtype,
+                    )
+            else:
+                raise RuntimeError(
+                    f"{spec} has no tensor metadata."
+                )
+
+        def _default_ret_value(arg):
+            type_str = str(arg.type)
+            if type_str == "()":  # used in aten function schema for None return value
+                return None
+
+            if _arg_type_is_primitive(type_str):
+                obj_list = [None for _ in range(dist.get_world_size())]
+                dist.all_gather_object(obj_list, None)
+                obj_list = list(filter(lambda x: x is not None, obj_list))
+                # TODO: this nested if-else statement is due to MYPY's
+                # issue on inferring reduce op's type. We have to explicitly
+                # type the lambda to make MYPY happy. But we better find a
+                # smarter way to do it.
+                if type_str == "bool":
+                    return functools.reduce(lambda x, y: x and y, obj_list, True)  # type: ignore[arg-type]
+                elif (
+                    (type_str == "int")
+                    or (type_str == "float")
+                    or (type_str == "double")
+                ):
+                    return functools.reduce(lambda x, y: x + y, obj_list, 0)  # type: ignore[arg-type]
+                else:
+                    raise NotImplementedError(
+                        f"Unsupported primitive type {type_str}."
+                    )
+
+            if type_str == "Tensor":
+                assert isinstance(output_sharding.output_spec, DTensorSpec), (
+                    f"output spec does not match with output! Expected "
+                    f"DTensorSpec, got {output_sharding.output_spec}"
+                )
+                return _default_tensor(output_sharding.output_spec)
+            elif type_str == "List[Tensor]":
+                assert isinstance(output_sharding.output_spec, Sequence), (
+                    f"output spec does not match with output! Expected "
+                    f"Sequence[DTensorSpec], got {output_sharding.output_spec}"
+                )
+                return [
+                    _default_tensor(spec)
```

Comment:
actually could we do this List logic in `default_tensor` function directly? maybe rename it to sth like `default_tensor_types`, and handle the list variant inside.

Reply:
The reason I extract `default_tensor` instead of handling Tensor and List[Tensor] in one function is, I want to avoid that this logic appears twice. Otherwise:
```
default_tensor_types(output_spec):
    if (is_tensor()):
        extract dtype
        return torch.tensor(...)
    elif (is_tensor_list()):
        result = []
        for spec in output_spec:
            extract dtype
            result.append(torch.tensor(...))
        return result
```

But we can split `default_ret_value` into `default_primitive_types` and `default_tensor_types`. And in `default_tensor_types` we check it's a Tensor or List[Tensor]. For List[Tensor], we do the list comprehension and call `default_tensor`; for Tensor, call `default_tensor`.  What do you think?

Reply:
I think the current way of structing the code is too messy, we should reduce the number of util functions to make it more modular, and unify some check logics if possible.